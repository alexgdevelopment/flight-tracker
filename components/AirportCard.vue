<template>
  <v-card elevation="5">
    <v-list-item
      ><v-list-item-content
        ><v-list-item-title data-testid="airportName">{{
          airport.fullName
        }}</v-list-item-title
        ><v-list-item-subtitle
          >Airport name</v-list-item-subtitle
        ></v-list-item-content
      ></v-list-item
    >
    <v-list-item
      ><v-list-item-content
        ><v-list-item-title data-testid="airportCountryName">{{
          airport.country.name
        }}</v-list-item-title
        ><v-list-item-subtitle
          >Country</v-list-item-subtitle
        ></v-list-item-content
      ></v-list-item
    >
  </v-card>
</template>

<script>
  export default {
    name: 'AirportCard',
    props: {
      airport: {
        type: Object,
        required: true,
      },
    },
  }
</script>