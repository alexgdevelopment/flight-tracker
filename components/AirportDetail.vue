<template>
  <v-col>
    <v-card class="pa-3">
      <v-fab-transition>
        <v-btn
          :color="mainColor"
          elevation="2"
          absolute
          top
          left
          fab
          x-small
          data-testid="backButton"
          @click="onBack"
        >
          <v-icon>mdi-arrow-left</v-icon>
        </v-btn>
      </v-fab-transition>
      <AirportCard v-if="airport" ref="airportCard" :airport="airport" />
      <v-row justify="center" class="mt-3">
        <v-col>
          <v-card elevation="5" class="pa-3">
            <v-row justify="center">
              <v-col cols="auto">
                <div>From</div>
              </v-col>
            </v-row>
            <v-menu
              v-model="dateFromPicker"
              :close-on-content-click="false"
              offset-y
            >
              <template #activator="{ on, attrs }">
                <v-text-field
                  v-model="dateFrom"
                  readonly
                  prepend-icon="mdi-calendar"
                  data-testid="dateFromTextField"
                  v-bind="attrs"
                  v-on="on"
                >
                </v-text-field>
              </template>
              <v-date-picker
                v-model="dateFrom"
                :color="mainColor"
                :max="dateTo"
                :min="oneDayLess(dateTo)"
                data-testid="dateFromPicker"
                @click:date="dateFromPicker = false"
              />
            </v-menu>
            <v-menu
              v-model="timeFromPicker"
              :close-on-content-click="false"
              offset-y
            >
              <template #activator="{ on, attrs }">
                <v-text-field
                  v-model="timeFrom"
                  readonly
                  prepend-icon="mdi-clock"
                  data-testid="timeFromTextField"
                  v-bind="attrs"
                  v-on="on"
                >
                </v-text-field>
              </template>
              <v-time-picker
                v-model="timeFrom"
                :color="mainColor"
                format="24hr"
                :min="twelveHoursLess(timeTo)"
                data-testid="timeFromPicker"
                @click:minute="timeFromPicker = false"
              />
            </v-menu>
          </v-card>
        </v-col>
        <v-col>
          <v-card elevation="5" class="pa-3">
            <v-row justify="center">
              <v-col cols="auto">
                <div>To</div>
              </v-col>
            </v-row>
            <v-menu
              v-model="dateToPicker"
              :close-on-content-click="false"
              offset-y
            >
              <template #activator="{ on, attrs }">
                <v-text-field
                  v-model="dateTo"
                  readonly
                  prepend-icon="mdi-calendar"
                  data-testid="dateToTextField"
                  v-bind="attrs"
                  v-on="on"
                >
                </v-text-field>
              </template>
              <v-date-picker
                v-model="dateTo"
                :color="mainColor"
                :min="dateFrom"
                :max="oneDayMore(dateFrom)"
                data-testid="dateToPicker"
                @click:date="dateToPicker = false"
              />
            </v-menu>
            <v-menu
              v-model="timeToPicker"
              :close-on-content-click="false"
              offset-y
            >
              <template #activator="{ on, attrs }">
                <v-text-field
                  v-model="timeTo"
                  readonly
                  prepend-icon="mdi-clock"
                  data-testid="timeToTextField"
                  v-bind="attrs"
                  v-on="on"
                >
                </v-text-field>
              </template>
              <v-time-picker
                v-model="timeTo"
                :color="mainColor"
                :max="twelveHoursMore(timeFrom)"
                format="24hr"
                data-testid="timeToPicker"
                @click:minute="timeToPicker = false"
              />
            </v-menu>
          </v-card>
        </v-col>
      </v-row>
      <v-row justify="center" class="mt-6">
        <v-btn data-testid="clearTermsButton" @click.native="onClear">
          clear terms
        </v-btn>
        <v-btn data-testid="searchButton" @click.native="onSearch">
          search
        </v-btn>
      </v-row>
      <v-row justify="center">
        <v-col
          v-if="arrivals && arrivals.length !== 0"
          data-testid="arrivalsColumn"
        >
          <v-row justify="center">
            <v-col cols="auto">
              <div>Arrivals</div>
            </v-col>
          </v-row>
          <FlightCard
            v-for="(flight, index) in arrivals"
            :key="index"
            :flight="flight"
          />
        </v-col>
        <v-col
          v-if="departures && departures.length !== 0"
          data-testid="departuresColumn"
          ><v-row justify="center">
            <v-col cols="auto">
              <div>Departures</div>
            </v-col>
          </v-row>
          <FlightCard
            v-for="(flight, index) in departures"
            :key="index"
            :flight="flight"
          />
        </v-col>
      </v-row>
    </v-card>
  </v-col>
</template>

<script>
  import { mapState } from 'vuex'

  export default {
    data() {
      return {
        dateFromPicker: false,
        timeFromPicker: false,
        dateToPicker: false,
        timeToPicker: false,
      }
    },
    computed: {
      ...mapState({
        airport: (state) => state.airports.currentAirport,
        arrivals: (state) => state.flights.arrivals,
        departures: (state) => state.flights.departures,
        mainColor: (state) => state.mainColor,
      }),
      dateFrom: {
        get() {
          return this.$store.state.dateFrom
        },
        set(value) {
          this.$store.commit('set', { varName: 'dateFrom', value })
        },
      },
      dateTo: {
        get() {
          return this.$store.state.dateTo
        },
        set(value) {
          this.$store.commit('set', { varName: 'dateTo', value })
        },
      },
      timeFrom: {
        get() {
          return this.$store.state.timeFrom
        },
        set(value) {
          this.$store.commit('set', { varName: 'timeFrom', value })
        },
      },
      timeTo: {
        get() {
          return this.$store.state.timeTo
        },
        set(value) {
          this.$store.commit('set', { varName: 'timeTo', value })
        },
      },
    },
    methods: {
      onSearch() {
        this.$store.dispatch('flights/getFlightsForAirport')
      },
      onClear() {
        this.$store.dispatch('clearToFromDates')
      },
      onBack() {
        this.$store.dispatch('clearCurrentAirport')
      },
      oneDayLess(date) {
        if (!date) return

        let newDate = new Date(date)
        newDate = newDate.getTime() - 1000 * 60 * 60 * 24
        return new Date(newDate).toISOString().substring(0, 10)
      },
      oneDayMore(date) {
        if (!date) return
        let newDate = new Date(date)
        newDate = newDate.getTime() + 1000 * 60 * 60 * 24
        return new Date(newDate).toISOString().substring(0, 10)
      },
      twelveHoursMore(date) {
        if (!date) return
        let newDate = new Date()
        const dateArr = date.split(':')
        newDate.setHours(+dateArr[0], +dateArr[1])
        newDate = newDate.getTime() + 1000 * 60 * 60 * 12
        newDate = new Date(newDate)
        let hours = newDate.getHours()
        if (hours < 10) {
          hours = `0${hours}`
        }
        let minutes = newDate.getMinutes()
        if (minutes < 10) {
          minutes = `0${minutes}`
        }
        return `${hours}:${minutes}`
      },
      twelveHoursLess(date) {
        if (!date) return
        let newDate = new Date()
        const dateArr = date.split(':')
        newDate.setHours(+dateArr[0], +dateArr[1])
        newDate = newDate.getTime() - 1000 * 60 * 60 * 12
        newDate = new Date(newDate)
        let hours = newDate.getHours()
        if (hours < 10) {
          hours = `0${hours}`
        }
        let minutes = newDate.getMinutes()
        if (minutes < 10) {
          minutes = `0${minutes}`
        }
        return `${hours}:${minutes}`
      },
    },
  }
</script>