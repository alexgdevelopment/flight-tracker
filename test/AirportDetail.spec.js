import { mount } from '@vue/test-utils'
import { Store } from 'vuex'
import Vuetify from 'vuetify'
import { addElemWithDataAppToBody } from './helpers'
import { mockAirport, mockFlight } from './mockObjects'
import AirportDetail from '@/components/AirportDetail.vue'
import AirportCard from '@/components/AirportCard.vue'
describe('Airport Detail', () => {
  let wrapper = null
  let state
  let store
  let vuetify
  let mutations
  let data
  let actions

  addElemWithDataAppToBody()

  beforeEach(() => {
    vuetify = new Vuetify()
    state = {
      mainColor: 'green',
    }
    mutations = {
      set: jest.fn(),
    }
    actions = {
      'flights/getFlightsForAirport': jest.fn(),
      clearToFromDates: jest.fn(),
      clearCurrentAirport: jest.fn(),
    }
    store = new Store({
      state,
      mutations,
      actions,
      modules: {
        airports: {
          state: {
            currentAirport: mockAirport,
          },
        },
        flights: {
          state: {
            arrivals: [mockFlight],
            departures: [mockFlight],
          },
        },
      },
    })
    data = () => ({
      dateFromPicker: false,
      timeFromPicker: false,
      dateToPicker: false,
      timeToPicker: false,
    })
    wrapper = mount(AirportDetail, {
      vuetify,
      stubs: {
        AirportCard: {
          template: '<div class="airport-card" />',
        },
        'v-time-picker': {
          template:
            '<div class="v-time-picker" @click="$store.commit(`set`)" />',
        },
        FlightCard: {
          template: '<div class="flight-card" />',
        },
      },
      data,
      store,
    })
  })
  test('is a Vue instance', () => {
    expect(wrapper.vm).toBeTruthy()
  })

  it('should match the snapshot', () => {
    expect(wrapper.element).toMatchSnapshot()
  })

  test('should show an airport card', () => {
    expect(wrapper.findComponent(AirportCard)).toBeTruthy()
  })

  test('should emit action on dateFrom change', async () => {
    await wrapper.find('[data-testid="dateFromTextField"').trigger('click')
    const dateFromPicker = wrapper.find('[data-testid="dateFromPicker"]')

    await dateFromPicker.find('tr > td > button').trigger('click')
    expect(mutations.set).toHaveBeenCalled()
  })
  test('should emit correct action on dateTo change', async () => {
    await wrapper.find('[data-testid="dateToTextField"').trigger('click')
    const dateToPicker = wrapper.find('[data-testid="dateToPicker"]')

    await dateToPicker.find('tr > td > button').trigger('click')
    expect(mutations.set).toHaveBeenCalled()
  })
  test('should emit correct action on timeFrom change', async () => {
    await wrapper.find('[data-testid="timeFromTextField"').trigger('click')
    await wrapper.find('[data-testid="timeFromPicker"]').trigger('click')
    expect(mutations.set).toHaveBeenCalled()
  })
  test('should emit correct action on timeTo change', async () => {
    await wrapper.find('[data-testid="timeToTextField"').trigger('click')
    await wrapper.find('[data-testid="timeToPicker"]').trigger('click')
    expect(mutations.set).toHaveBeenCalled()
  })

  test('should emit search action on search button click', async () => {
    await wrapper.find('[data-testid="searchButton"]').trigger('click')
    expect(actions['flights/getFlightsForAirport']).toHaveBeenCalled()
  })

  test('should emit correct action on clearDates click', () => {
    wrapper.find('[data-testid="clearTermsButton"]').trigger('click')
    expect(actions.clearToFromDates).toHaveBeenCalled()
  })

  test('should emit correct action on back click', () => {
    wrapper.find('[data-testid="backButton"]').trigger('click')
    expect(actions.clearCurrentAirport).toHaveBeenCalled()
  })

  test('should display arrivals column if there are any in the store', () => {
    expect(wrapper.find('[data-testid="arrivalsColumn"]')).toBeTruthy()
  })
  test('should display departures column if there are any in the store', () => {
    expect(wrapper.find('[data-testid="departuresColumn"]')).toBeTruthy()
  })
  test('should compute correct max and min dates', () => {
    expect(wrapper.vm.oneDayLess('2020-01-01')).toEqual('2019-12-31')
    expect(wrapper.vm.oneDayMore('2020-01-31')).toEqual('2020-02-01')
    expect(wrapper.vm.twelveHoursLess('04:00')).toEqual('16:00')
    expect(wrapper.vm.twelveHoursMore('13:00')).toEqual('01:00')
  })
})
