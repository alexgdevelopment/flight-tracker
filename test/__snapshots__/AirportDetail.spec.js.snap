// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Airport Detail should match the snapshot 1`] = `
<div
  class="col"
>
  <div
    class="pa-3 v-card v-sheet theme--light"
  >
    <transition-stub
      mode="out-in"
      name="fab-transition"
    >
      <button
        class="v-btn v-btn--absolute v-btn--is-elevated v-btn--fab v-btn--has-bg v-btn--left v-btn--round v-btn--top theme--light elevation-2 v-size--x-small green"
        data-testid="backButton"
        type="button"
      >
        <span
          class="v-btn__content"
        >
          <i
            aria-hidden="true"
            class="v-icon notranslate mdi mdi-arrow-left theme--light"
          />
        </span>
      </button>
    </transition-stub>
     
    <div
      airport="[object Object]"
      class="airport-card"
    />
     
    <div
      class="row mt-3 justify-center"
    >
      <div
        class="col"
      >
        <div
          class="pa-3 v-card v-sheet theme--light elevation-5"
        >
          <div
            class="row justify-center"
          >
            <div
              class="col col-auto"
            >
              <div>
                From
              </div>
            </div>
          </div>
           
          <div
            class="v-menu"
          >
            <!---->
          </div>
          <div
            class="v-input v-input--is-readonly theme--light v-text-field"
          >
            <div
              class="v-input__prepend-outer"
            >
              <div
                class="v-input__icon v-input__icon--prepend"
              >
                <i
                  aria-hidden="true"
                  class="v-icon notranslate mdi mdi-calendar theme--light"
                />
              </div>
            </div>
            <div
              class="v-input__control"
            >
              <div
                class="v-input__slot"
              >
                <div
                  class="v-text-field__slot"
                >
                  <input
                    aria-expanded="false"
                    aria-haspopup="true"
                    data-testid="dateFromTextField"
                    id="input-46"
                    readonly="readonly"
                    role="button"
                    type="text"
                  />
                </div>
              </div>
              <div
                class="v-text-field__details"
              >
                <div
                  class="v-messages theme--light"
                >
                  <transition-group-stub
                    class="v-messages__wrapper"
                    name="message-transition"
                    tag="div"
                  />
                </div>
              </div>
            </div>
          </div>
           
          <div
            class="v-menu"
          >
            <!---->
          </div>
          <div
            class="v-input v-input--is-readonly theme--light v-text-field"
          >
            <div
              class="v-input__prepend-outer"
            >
              <div
                class="v-input__icon v-input__icon--prepend"
              >
                <i
                  aria-hidden="true"
                  class="v-icon notranslate mdi mdi-clock theme--light"
                />
              </div>
            </div>
            <div
              class="v-input__control"
            >
              <div
                class="v-input__slot"
              >
                <div
                  class="v-text-field__slot"
                >
                  <input
                    aria-expanded="false"
                    aria-haspopup="true"
                    data-testid="timeFromTextField"
                    id="input-51"
                    readonly="readonly"
                    role="button"
                    type="text"
                  />
                </div>
              </div>
              <div
                class="v-text-field__details"
              >
                <div
                  class="v-messages theme--light"
                >
                  <transition-group-stub
                    class="v-messages__wrapper"
                    name="message-transition"
                    tag="div"
                  />
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
       
      <div
        class="col"
      >
        <div
          class="pa-3 v-card v-sheet theme--light elevation-5"
        >
          <div
            class="row justify-center"
          >
            <div
              class="col col-auto"
            >
              <div>
                To
              </div>
            </div>
          </div>
           
          <div
            class="v-menu"
          >
            <!---->
          </div>
          <div
            class="v-input v-input--is-readonly theme--light v-text-field"
          >
            <div
              class="v-input__prepend-outer"
            >
              <div
                class="v-input__icon v-input__icon--prepend"
              >
                <i
                  aria-hidden="true"
                  class="v-icon notranslate mdi mdi-calendar theme--light"
                />
              </div>
            </div>
            <div
              class="v-input__control"
            >
              <div
                class="v-input__slot"
              >
                <div
                  class="v-text-field__slot"
                >
                  <input
                    aria-expanded="false"
                    aria-haspopup="true"
                    data-testid="dateToTextField"
                    id="input-57"
                    readonly="readonly"
                    role="button"
                    type="text"
                  />
                </div>
              </div>
              <div
                class="v-text-field__details"
              >
                <div
                  class="v-messages theme--light"
                >
                  <transition-group-stub
                    class="v-messages__wrapper"
                    name="message-transition"
                    tag="div"
                  />
                </div>
              </div>
            </div>
          </div>
           
          <div
            class="v-menu"
          >
            <!---->
          </div>
          <div
            class="v-input v-input--is-readonly theme--light v-text-field"
          >
            <div
              class="v-input__prepend-outer"
            >
              <div
                class="v-input__icon v-input__icon--prepend"
              >
                <i
                  aria-hidden="true"
                  class="v-icon notranslate mdi mdi-clock theme--light"
                />
              </div>
            </div>
            <div
              class="v-input__control"
            >
              <div
                class="v-input__slot"
              >
                <div
                  class="v-text-field__slot"
                >
                  <input
                    aria-expanded="false"
                    aria-haspopup="true"
                    data-testid="timeToTextField"
                    id="input-62"
                    readonly="readonly"
                    role="button"
                    type="text"
                  />
                </div>
              </div>
              <div
                class="v-text-field__details"
              >
                <div
                  class="v-messages theme--light"
                >
                  <transition-group-stub
                    class="v-messages__wrapper"
                    name="message-transition"
                    tag="div"
                  />
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
     
    <div
      class="row mt-6 justify-center"
    >
      <button
        class="v-btn v-btn--is-elevated v-btn--has-bg theme--light v-size--default"
        data-testid="clearTermsButton"
        type="button"
      >
        <span
          class="v-btn__content"
        >
          
        clear terms
      
        </span>
      </button>
       
      <button
        class="v-btn v-btn--is-elevated v-btn--has-bg theme--light v-size--default"
        data-testid="searchButton"
        type="button"
      >
        <span
          class="v-btn__content"
        >
          
        search
      
        </span>
      </button>
    </div>
     
    <div
      class="row justify-center"
    >
      <div
        class="col"
        data-testid="arrivalsColumn"
      >
        <div
          class="row justify-center"
        >
          <div
            class="col col-auto"
          >
            <div>
              Arrivals
            </div>
          </div>
        </div>
         
        <div
          class="flight-card"
          flight="[object Object]"
        />
      </div>
       
      <div
        class="col"
        data-testid="departuresColumn"
      >
        <div
          class="row justify-center"
        >
          <div
            class="col col-auto"
          >
            <div>
              Departures
            </div>
          </div>
        </div>
         
        <div
          class="flight-card"
          flight="[object Object]"
        />
      </div>
    </div>
  </div>
</div>
`;
